File: miscellaneous/Miscellaneous/Tutorials/2009-EDSys/Examples/C#/SchedIntro.cs
// -------------------------------------------------------------- -*- C# -*-
// File: examples/src/csharp/SchedIntro.cs
// --------------------------------------------------------------------------
// Copyright (C) 1990-2008 by ILOG.
// All Rights Reserved.
//
// Permission is expressly granted to use this example in the
// course of developing applications that use ILOG products.
// --------------------------------------------------------------------------

/* ------------------------------------------------------------

Problem Description
-------------------

This is a basic problem that involves building a house. The masonry,
roofing, painting, etc.  must be scheduled. Some tasks must
necessarily take place before others, and these requirements are
expressed through precedence constraints.

------------------------------------------------------------ */

//$doc:ALL
//$doc:TASKS
using System;
using System.IO;
using ILOG.CP;
using ILOG.Concert;

namespace SchedIntro
{
    public class SchedIntro
    {

        static void Main()
        {
            CP cp = new CP();
            /// CREATE THE TIME-INTERVALS.///
            IIntervalVar masonry = cp.IntervalVar(35, "masonry   ");
            IIntervalVar carpentry = cp.IntervalVar(15, "carpentry ");
            IIntervalVar plumbing = cp.IntervalVar(40, "plumbing  ");
            IIntervalVar ceiling = cp.IntervalVar(15, "ceiling   ");
            IIntervalVar roofing = cp.IntervalVar(5, "roofing   ");
            IIntervalVar painting = cp.IntervalVar(10, "painting  ");
            IIntervalVar windows = cp.IntervalVar(5, "windows   ");
            IIntervalVar facade = cp.IntervalVar(10, "facade    ");
            IIntervalVar garden = cp.IntervalVar(5, "garden    ");
            IIntervalVar moving = cp.IntervalVar(5, "moving    ");
            //end:TASKS

            /// ADDING TEMPORAL CONSTRAINTS.///
            //$doc:CSTS
            cp.Add(cp.EndBeforeStart(masonry, carpentry));
            cp.Add(cp.EndBeforeStart(masonry, plumbing));
            cp.Add(cp.EndBeforeStart(masonry, ceiling));
            cp.Add(cp.EndBeforeStart(carpentry, roofing));
            cp.Add(cp.EndBeforeStart(ceiling, painting));
            cp.Add(cp.EndBeforeStart(roofing, windows));
            cp.Add(cp.EndBeforeStart(roofing, facade));
            cp.Add(cp.EndBeforeStart(plumbing, facade));
            cp.Add(cp.EndBeforeStart(roofing, garden));
            cp.Add(cp.EndBeforeStart(plumbing, garden));
            cp.Add(cp.EndBeforeStart(windows, moving));
            cp.Add(cp.EndBeforeStart(facade, moving));
            cp.Add(cp.EndBeforeStart(garden, moving));
            cp.Add(cp.EndBeforeStart(painting, moving));
            //end:CSTS

            /// EXTRACTING THE MODEL AND SOLVING.///
            //$doc:SOLVE
            if (cp.Solve())
            {
                Console.WriteLine(cp.GetDomain(masonry));
                Console.WriteLine(cp.GetDomain(carpentry));
                Console.WriteLine(cp.GetDomain(plumbing));
                Console.WriteLine(cp.GetDomain(ceiling));
                Console.WriteLine(cp.GetDomain(roofing));
                Console.WriteLine(cp.GetDomain(painting));
                Console.WriteLine(cp.GetDomain(windows));
                Console.WriteLine(cp.GetDomain(facade));
                Console.WriteLine(cp.GetDomain(garden));
                Console.WriteLine(cp.GetDomain(moving));
            }
            else
            {
                Console.Write("No solution found. ");
            }
            cp.PrintInformation();
        }
    }
}
//end:SOLVE
//end:ALL

File: miscellaneous/Miscellaneous/Tutorials/2009-EDSys/Examples/Java/SchedIntro.java
// ------------------------------------------------------------- -*- Java -*-
// File: examples/src/java/SchedIntro.java
// --------------------------------------------------------------------------
// Copyright (C) 1990-2008 by ILOG.
// All Rights Reserved.
//
// Permission is expressly granted to use this example in the
// course of developing applications that use ILOG products.
// --------------------------------------------------------------------------

/* ------------------------------------------------------------

Problem Description
-------------------

This is a basic problem that involves building a house. The masonry,
roofing, painting, etc.  must be scheduled. Some tasks must
necessarily take place before others, and these requirements are
expressed through precedence constraints.

------------------------------------------------------------ */

//$doc:ALL
//$doc:TASKS
import ilog.concert.*;
import ilog.cp.*;

public class SchedIntro {

    public static void main(String[] args) {
        try {
            IloCP cp = new IloCP();
            
            /* CREATE THE INTERVAL VARIABLES. */
            IloIntervalVar masonry   = cp.intervalVar(35, "masonry   ");
            IloIntervalVar carpentry = cp.intervalVar(15, "carpentry ");
            IloIntervalVar plumbing  = cp.intervalVar(40, "plumbing  ");
            IloIntervalVar ceiling   = cp.intervalVar(15, "ceiling   ");
            IloIntervalVar roofing   = cp.intervalVar(5,  "roofing   ");
            IloIntervalVar painting  = cp.intervalVar(10, "painting  ");
            IloIntervalVar windows   = cp.intervalVar(5,  "windows   ");
            IloIntervalVar facade    = cp.intervalVar(10, "facade    ");
            IloIntervalVar garden    = cp.intervalVar(5,  "garden    ");
            IloIntervalVar moving    = cp.intervalVar(5,  "moving    ");
//end:TASKS

            /* ADDING PRECEDENCE CONSTRAINTS. */
//$doc:CSTS
            cp.add(cp.endBeforeStart(masonry,   carpentry));
            cp.add(cp.endBeforeStart(masonry,   plumbing));
            cp.add(cp.endBeforeStart(masonry,   ceiling));
            cp.add(cp.endBeforeStart(carpentry, roofing));
            cp.add(cp.endBeforeStart(ceiling,   painting));
            cp.add(cp.endBeforeStart(roofing,   windows));
            cp.add(cp.endBeforeStart(roofing,   facade));
            cp.add(cp.endBeforeStart(plumbing,  facade));
            cp.add(cp.endBeforeStart(roofing,   garden));
            cp.add(cp.endBeforeStart(plumbing,  garden));
            cp.add(cp.endBeforeStart(windows,   moving));
            cp.add(cp.endBeforeStart(facade,    moving));
            cp.add(cp.endBeforeStart(garden,    moving));
            cp.add(cp.endBeforeStart(painting,  moving));
//end:CSTS

            /* EXTRACTING THE MODEL AND SOLVING. */
//$doc:SOLVE
            if (cp.solve()) {
                System.out.println(cp.getDomain(masonry));
                System.out.println(cp.getDomain(carpentry));
                System.out.println(cp.getDomain(plumbing));
                System.out.println(cp.getDomain(ceiling));
                System.out.println(cp.getDomain(roofing));
                System.out.println(cp.getDomain(painting));
                System.out.println(cp.getDomain(windows));
                System.out.println(cp.getDomain(facade));
                System.out.println(cp.getDomain(garden));
                System.out.println(cp.getDomain(moving));
            } else {
                System.out.print("No solution found. ");
            }
            cp.printInformation();
        } catch (IloException e) {
            System.err.println("Error " + e );
        }
    }
}
//end:SOLVE
//end:ALL